### DISCUSSION TITLE
`MACOSX_DEPLOYMENT_TARGET` ignored?

### DISCUSSION THREAD

**Discussion by daemontus** [PyO3/maturin#1342] [Q&A] [answered]
Hi!

I'm building a library that depends on the `z3-sys` crate which internally uses the Z3 theorem prover. The problem is that `z3-sys` can link Z3 statically (which I want: my python tool is rather standalone and I don't want to force people to hunt for extra dependencies). However, this means that I'm building whole Z3 alongside my crate. This is not an issue on linux or windows, but on macos, Z3 uses some APIs that are only available on newer macOS versions (e.g. 11.0, whereas default "minimum" for rust is 10.8 I believe). 

Normally, you set this limit using `MACOSX_DEPLOYMENT_TARGET` environment variable, which should propagate to `cmake` that is actually building Z3. However, no matter how I try to set this variable, the build process for Z3 does not seem to notice or care and I get API compatibility errors (see for example [here](https://github.com/daemontus/biodivine-aeon-py/actions/runs/3638122471)). 

Interestingly, on my local macOS laptop with macOS 13, the problem does not seem to be occurring. That is, I don't even have to set `MACOSX_DEPLOYMENT_TARGET` and everything works. However, so far I haven't found a single difference between my local setup and the github actions runner that could explain this.

Does anyone have any idea how to resolve this? Have you encountered something similar? (I guess I could try to fork `z3-sys` and force the variable in the build script, but I'd like to avoid that if possible)

**Comment by messense** [comment]
In maturin we only use `MACOSX_DEPLOYMENT_TARGET` in two places

1. https://github.com/PyO3/maturin/blob/82aa66a596ca58a1771188a7362aea884b7424cc/src/target.rs#L294 to calculate wheel tags, doesn't affect `cargo build`.
2. https://github.com/PyO3/maturin/blob/82aa66a596ca58a1771188a7362aea884b7424cc/src/compile.rs#L402-L407 set default `MACOSX_DEPLOYMENT_TARGET` when the env var isn't set by user which in your case won't be used since you already set it.

so in conclusion maturin doesn't ignore `MACOSX_DEPLOYMENT_TARGET`, `cargo rustc` will inherit the env var you set normally.

> echo "MACOSX_DEPLOYMENT_TARGET=10.16" >> $GITHUB_ENV

Maybe try `11.0`? I don't think there is a macOS 10.16 release.

**Reply by daemontus** [reply]
Thank you for the suggestions! I tried `11.0` later (e.g. [here](https://github.com/daemontus/biodivine-aeon-py/actions/runs/3638358062/jobs/6140464925)), but it didn't do anything. 


But I think I figured it out! Based on [this](https://cmake.org/cmake/help/latest/envvar/MACOSX_DEPLOYMENT_TARGET.html), I tried setting `CMAKE_OSX_DEPLOYMENT_TARGET` to `11.0` and that actually seems to resolve the problem. So, either I don't understand how `cmakes` propagates this information (because based on the documentation above, I would assume `MACOSX_DEPLOYMENT_TARGET ` is read from the environment by `cmake`), or the github runner has some version of `cmake` where this behaves differently?

Regardless... it seems to be working. Thank you for your help!

**Comment by daemontus** [comment, marked as answer]
Just to close this with an actual answer...

For whatever reason, `MACOSX_DEPLOYMENT_TARGET ` did not do anything, but setting `CMAKE_OSX_DEPLOYMENT_TARGET` seems to resolve this.

Edit: I stand corrected. It turns out that **both** variables need to be set for the build to work.

**Reply by messense** [reply]
Turns out there was a hardcoded `MACOSX_DEPLOYMENT_TARGET = 10.9` in `maturin-action`: https://github.com/PyO3/maturin-action/issues/112, it might have caused issue for universal2/aarch64 build, fixed now.

**Reply by daemontus** [reply]
![surprise pikachu face](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2F1.bp.blogspot.com%2F-LfehkrZidUw%2FXheBc_gzEYI%2FAAAAAAAAM8w%2Fy9nKq-9YpB8m5LfUmHdb9fCHKBLY8BMxwCLcBGAsYHQ%2Fs200%2Fstandard-surprised-pikachu.png&f=1&nofb=1&ipt=dd614cd2994658af5f418e703440560865ffc1e3b70813dd7da5bd313940397b&ipo=images)
